### ISSUE TITLE
PeftModel trainable parameters with multiple adapters

### ISSUE THREAD

**Issue by NguyenRichard** [huggingface/peft#2759]
### System Info

peft-0.17.1
python 3.9

### Who can help?

@BenjaminBossan 

### Reproduction

**1) modules_to_save gradient true even when is_trainable=False**

The adapters has both modules_to_save and target_modules

```
peft_backbone = PeftModel.from_pretrained(
                    target_backbone,
                    safe_encoder_adapter_path1,
                    adapter_name=adapter_name1,
                    is_trainable=False
                )
 status = peft_backbone.get_model_status()
 check_trainable_params(target_backbone)
```

```
def check_trainable_params(model, print_layers=True):
    total_params = 0
    trainable_params = 0
    for name, param in model.named_parameters():
        num_params = param.numel()
        total_params += num_params
        if param.requires_grad:
            trainable_params += num_params
            if print_layers:
                print(f"[TRAINABLE] {name} - shape: {tuple(param.shape)}")
        elif print_layers:
            print(f"[FROZEN]    {name} - shape: {tuple(param.shape)}")

    print(f"\nTotal parameters: {total_params:,}")
    print(f"Trainable parameters: {trainable_params:,}")
    print(f"Frozen parameters: {total_params - trainable_params:,}")
    print(f"Trainable ratio: {100 * trainable_params / total_params:.2f}%")

    return trainable_params, total_params
```

example of printed trainable params 
[TRAINABLE] blocks.0.modules_to_save.adapter1.norm1.weight - shape: (1408,)
[FROZEN]    blocks.2.attn.qkv.lora_A.adapter1.weight - shape: (32, 1408)

**2) Loading an adapter after using from_pretrained**
```
peft_backbone = PeftModel.from_pretrained(
    target_backbone,
    safe_encoder_adapter_path1,
    adapter_name=modality_name,
    is_trainable=False
)
status = peft_backbone.get_model_status()
target_backbone.load_adapter(safe_encoder_adapter_path2, is_trainable=False, adapter_name=adapter2)
status = peft_backbone.get_model_status()
```

status before load_adapter shows {'adapter1': False} while after the load_adapter {'adapter2': False, 'adapter1': True}

I think the issue comes from BaseTurnerLayer.set_adapter that set True all my adapter1 lora layers' gradient while setting properly the adapter2 lora layers' gradient to False.
BaseTurnerLayer.set_adapter is called when doing self.add_adapter in PeftModel.load_adapter.

### Expected behavior

**1) modules_to_save gradient true even when is_trainable=False**

Expecting the gradients for modules_to_save layers to be false. It's working properly for lora layers.

**2) Loading an adapter after using from_pretrained**

Expecting adapter1 to remain gradient false (is_trainable=False during from_pretrained loading) even after loading another adapter.

**Other informations:**

Regarding issue 1), in the code of 2), the modules_to_save for adapter2 were properly set to false when using load_adapter with is_trainable=false.

[TRAINABLE] base_model.model.blocks.39.modules_to_save.adapter1.mlp.fc2.bias - shape: (1408,)
[FROZEN]    base_model.model.blocks.39.modules_to_save.adapter2.norm1.weight - shape: (1408,)

More generally, is there any reason peftmodel has to change the requires_gradient of adapters when calling set_adapter? (https://github.com/huggingface/peft/issues/2749)
I assume that it might be related to the fact that there might be a problem to have non activated adapter but with requires_gradient=True?
When using the library I was expecting to be able to set what params needed to be trained on all my adapters upon loading them with from_pretrained and load_adapter (or manually) then simply switch between adapters during the training with set_adapter.

**Comment by BenjaminBossan** [comment]
Thanks for reporting, I'm looking into this.

**Comment by BenjaminBossan** [comment]
@NguyenRichard I think the issue should be fixed with #2765. If you have time, could you please test that branch?

**Comment by NguyenRichard** [comment]
> [@NguyenRichard](https://github.com/NguyenRichard) I think the issue should be fixed with [#2765](https://github.com/huggingface/peft/pull/2765). If you have time, could you please test that branch?

I tested the branch on my code, the two issues are solved!

As you remarked in the fix, in my use case with frozen adapters and trained finetuned adapters, I still have to manually set the gradient after set_adapter due to the internal coupling even after this fix. 
But at least the instantiation is working properly now!

Should I close this issue?

**Comment by BenjaminBossan** [comment]
> I tested the branch on my code, the two issues are solved!

Nice.

> in my use case with frozen adapters and trained finetuned adapters, I still have to manually set the gradient after set_adapter due to the internal coupling even after this fix.

Could you show me the code for this? Just to be sure that it's not something else I haven't thought about.

> Should I close this issue?

Let's leave it open until the PR is merged.

**Comment by NguyenRichard** [comment]
> > I tested the branch on my code, the two issues are solved!
> 
> Nice.
> 
> > in my use case with frozen adapters and trained finetuned adapters, I still have to manually set the gradient after set_adapter due to the internal coupling even after this fix.
> 
> Could you show me the code for this? Just to be sure that it's not something else I haven't thought about.
> 
> > Should I close this issue?
> 
> Let's leave it open until the PR is merged.


To summarize how I use PEFT, I'm simply loading my 2 pretrained adapters, then add two new adapters that are used to finetune them
```
peft_backbone = PeftModel.from_pretrained(
                    target_backbone,
                    safe_encoder_adapter_path1,
                    adapter_name=adapter_name1,
                    is_trainable=False
                ) # lora and modules_to_save
target_backbone.load_adapter(safe_encoder_adapter_path2, is_trainable=False, adapter_name=adapter2) # lora and modules_to_save
target_backbone.add_adapter(ft_adapter1, lora_config) # lora only
target_backbone.add_adapter(ft_adapter2, lora_config) # lora only
```
At this stage, everything is properly set up as the first two adapters are non trainable and the latter two have require_gradient=True

in the forward function
```
target_backbone.base_model.set_adapter([adapter1, ft_adapter1])
status = backbone.get_model_status()
check_trainable_params(backbone)
```
This result in the {'adapter1': True, 'ft_adapter1': True, 'adapter2': False, 'ft_adapter2': False} and adapter1 requires gradient True for both lora and modules_to_save and ft_adapter1 requires radient True for it's only lora layers. adapter2 and ft_adapter2 have their gradient properly set to False

It makes sense as the code at the moment is intended to have set_adapter force the gradient to be either true or false regardless of its state upon loading.

I wonder if it would be reasonable to have a 'third option' for the inference_mode boolean, where the function 'do not change the gradient' of the adapter in the set_adapter?

**Comment by NguyenRichard** [comment]
Actually, I was wrong. I just checked the parameters after adding the ft_adapters (I forgot before), and there is a weird result.
```
peft_backbone = PeftModel.from_pretrained(
                    target_backbone,
                    safe_encoder_adapter_path1,
                    adapter_name=adapter_name1,
                    is_trainable=False
                ) # lora and modules_to_save
target_backbone.load_adapter(safe_encoder_adapter_path2, is_trainable=False, adapter_name=adapter2) # lora and modules_to_save
check_trainable_params(self.modality_to_backbone_map[mod]) # CHECK 1
status = self.modality_to_backbone_map[mod].get_model_status()
target_backbone.add_adapter(ft_adapter1, lora_config)
check_trainable_params(self.modality_to_backbone_map[mod]) # CHECK 2
status = self.modality_to_backbone_map[mod].get_model_status()
```

CHECK1: every weight false and requires_gradient false for both adapters as solved by your fix

CHECK2: adapter1 requires_gradient becomes unexpectedly True in get_model_status(). Moreover, only its lora layers gets their gradient to turn True and not the modules_to_save.
Another weird point is that the added adapter using add_adapter() is by default non trainable (I suppose it's because it's assumed that it will become trainable upon using set_adapter) requires_grade = {'adapter_2': False, 'adapter_1': True, 'ft_adapter1': False}

**Comment by BenjaminBossan** [comment]
Thanks for describing this yet uncovered bug. I can reproduce it and I [added it as x-failing](https://github.com/huggingface/peft/pull/2765/commits/870063b908a67a24d93bf94b230752de142e9ed4) to the PR.

> Another weird point is that the added adapter using add_adapter() is by default non trainable (I suppose it's because it's assumed that it will become trainable upon using set_adapter)

This is on purpose, the idea is that:

1. Calling `get_peft_model` or `PeftModel.from_pretrained` adds/loads an "active" adapter.
2. Calling `model.add_adapter` or `model.load_adapter` adds/loads an "inactive" adapter that needs to be activated with `model.set_adapter`.

> I wonder if it would be reasonable to have a 'third option' for the inference_mode boolean, where the function 'do not change the gradient' of the adapter in the set_adapter?

I think for a proper, clean solution, we should indeed completely decouple `set_adapter` and setting of `requires_grad`. Most of the time, they go hand in hand, but not always. This may require some breaking changes for some fairly common use cases though, so it's not an easy change to make.
